src/libdrakvuf/win-processes.c

```diff
@@ -1202,6 +1269,317 @@ bool win_find_mmvad(drakvuf_t drakvuf, addr_t eprocess, addr_t vaddr, mmvad_info
+    ++depth;
+
+
+    if (is_win7)
+    {
+        ctx.addr = node_addr + drakvuf->offsets[MMVAD_LEFT_CHILD];
+    }
+    else
+    {
+        ctx.addr = node_addr + drakvuf->offsets[RTL_BALANCED_NODE_LEFT];
+    }
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &left_child) != VMI_SUCCESS)
+    {
+        return false;
+    }
+
+    if (is_win7)
+    {
+        ctx.addr = node_addr + drakvuf->offsets[MMVAD_RIGHT_CHILD];
+    }
+    else
+    {
+        ctx.addr = node_addr + drakvuf->offsets[RTL_BALANCED_NODE_RIGHT];
+    }
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &right_child) != VMI_SUCCESS)
+    {
+        return false;
+    }
+
+    if (is_win7)
+    {
+        ctx.addr = node_addr + drakvuf->offsets[MMVAD_STARTING_VPN];
+
+        if (!is32bit)
+        {
+            if (vmi_read_64(drakvuf->vmi, &ctx, &starting_vpn) != VMI_SUCCESS)
+            {
+                return false;
+            }
+
+            ctx.addr = node_addr + drakvuf->offsets[MMVAD_ENDING_VPN];
+
+            if (vmi_read_64(drakvuf->vmi, &ctx, &ending_vpn) != VMI_SUCCESS)
+            {
+                return false;
+            }
+        }
+        else
+        {
+            if (vmi_read_32(drakvuf->vmi, &ctx, (uint32_t*)&starting_vpn) != VMI_SUCCESS)
+            {
+                return false;
+            }
+
+            ctx.addr = node_addr + drakvuf->offsets[MMVAD_ENDING_VPN];
+
+            if (vmi_read_32(drakvuf->vmi, &ctx, (uint32_t*)&ending_vpn) != VMI_SUCCESS)
+            {
+                return false;
+            }
+        }
+    }
+    else
+    {
+        mmvad_core = node_addr + drakvuf->offsets[MMVAD_CORE];
+        ctx.addr = mmvad_core + drakvuf->offsets[MMVAD_SHORT_STARTING_VPN];
+
+        if (vmi_read_32(drakvuf->vmi, &ctx, &starting_vpn_low) != VMI_SUCCESS)
+        {
+            return false;
+        }
+
+        ctx.addr = mmvad_core + drakvuf->offsets[MMVAD_SHORT_ENDING_VPN];
+
+        if (vmi_read_32(drakvuf->vmi, &ctx, &ending_vpn_low) != VMI_SUCCESS)
+        {
+            return false;
+        }
+
+        ctx.addr = mmvad_core + drakvuf->offsets[MMVAD_SHORT_STARTING_VPN_HIGH];
+
+        if (vmi_read_8(drakvuf->vmi, &ctx, &starting_vpn_high) != VMI_SUCCESS)
+        {
+            return false;
+        }
+
+        ctx.addr = mmvad_core + drakvuf->offsets[MMVAD_SHORT_ENDING_VPN_HIGH];
+
+        if (vmi_read_8(drakvuf->vmi, &ctx, &ending_vpn_high) != VMI_SUCCESS)
+        {
+            return false;
+        }
+
+        starting_vpn = starting_vpn_high;
+        starting_vpn <<= 32;
+        starting_vpn |= starting_vpn_low;
+
+        ending_vpn = ending_vpn_high;
+        ending_vpn <<= 32;
+        ending_vpn |= ending_vpn_low;
+    }
+
+    if (starting_vpn == 0 && ending_vpn == 0)
+    {
+        // the root node seems to be empty with only right child pointer filled in
+        return win_traverse_mmvad_node(drakvuf, right_child, callback, callback_data);
+    }
+
+    addr_t pool_tag_delta = is32bit ? 4 : 0xC;
```

Comment:
Seems to me it would be easier to read this if this function was split into two, one for the part that's above, and then the another for the code below from here.

Reply:
Also, use the defines here as well

Reply:
Done